File: 015.3Sum/sum.cc
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(), nums.end());

        vector<vector<int>> ans;
        for (int i = 0; i < int(nums.size()) - 2; i++) {
            int l = i + 1, h = nums.size() - 1;

            if (i != 0 && nums[i] == nums[i-1]) {
                continue;
            }

            while (l < h) {
                int part = nums[l] + nums[h] + nums[i];
                if (part == 0) {
                    vector<int> one = {nums[i], nums[l], nums[h]};
                    ans.push_back(one);

                    while (l < h && nums[l] == nums[l+1]) l++;
                    while (l < h && nums[h] == nums[h-1]) h--;
                    l++; h--;
                }

                else if (part < 0) l++;
                else h--;
            }
        }

        return ans;
    }
};

File: 015.3Sum/sum.py
class Solution(object):
    # Time limit exceeded
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        ans = []
        unique = set()
        for i in range(len(nums) - 2):
            s = -nums[i]
            m = {}
            for j in range(i+1, len(nums)):
                if nums[j] in m:
                    solution = [nums[i], -nums[i] - nums[j], nums[j]]
                    sorted_solution = tuple(sorted(solution))
                    if sorted_solution not in unique:
                        unique.add(sorted_solution)
                        ans.append(solution)
                    m[nums[j]] -= 1
                else:
                    m[s - nums[j]] = 1

        return ans


    def threeSum2(self, nums):
        nums.sort()
        ans = []
        for i in range(len(nums)):
            if i != 0 and nums[i] == nums[i-1]:
                continue
            s = -nums[i]
            l, h = i + 1, len(nums) - 1
            while l < h:
                if nums[l] + nums[h] == s:
                    ans.append([nums[i], nums[l], nums[h]])
                    while l < h and nums[l+1] == nums[l]:
                        l += 1
                    l += 1
                    while l < h and nums[h-1] == nums[h]:
                        h -= 1
                    h -= 1
                elif nums[l] + nums[h] < s:
                    l += 1
                else:
                    h -= 1
        return ans

if __name__ == '__main__':
    nums = [-1, -1, -1, 0, 1, 2, 2, 2, -1, 4]
    print Solution().threeSum2(nums)
